### ISSUE TITLE
Add error message to Interactor#fail! ?

### ISSUE THREAD

**Issue by pascalbetz** [hanami/utils#182]
Would it make sense to add an (optional) error message to `Hanami::Interactor#fail!`
?
Message could be handled like `Hanami::Interactor#error!(msg)`

**Comment by jodosha** [comment]
@pascalbetz `#fail!` is designed to just terminate the flow, while `#error!` is designed to store an error message **and** terminate the flow.

By adding that responsibility to `#fail!`, the semantic will be identical. But we still need a method that is just able to terminate the flow.

---

Let me explain better the differences.

When you have a situation where you want to add an error and exit, you can use `#error!("ouch")`. As said before, this terminates the flow, and reports the error message.

But when you want to report multiple error messages things are quite different. Let's pretend we have a **third-party library** (**NOT** from Hanami) that validates data for each attribute.

```ruby
def validate_my_data
  validator.attributes.each do |attr|
    error(attr.error) unless attr.error.nil?
  end

  fail! unless errors.empty?
end
```

We want to **accumulate** first all the errors, and then terminate the flow, **without** adding any extra error.

---

TL;DR: `#fail!` is designed to act only one level of abstraction: termination of flow. While `#error!` is an abstraction on top of `#fail!` which, add on error **and** terminates the flow.

Does my explanation solves your doubts?

**Comment by pascalbetz** [comment]
thx for the explanation...
should have checked the source code, did not know that `error!` exists. 

To me the interface would be easier if it were like
```ruby
def fail!(msg = nil)
  error(msg) if msg
  @__result.fail!
  throw :fail
end
```

**Comment by mereghost** [comment]
TBH I'm not sure I like `error!` as it does 2 unrelated things (add an error message and acts as flow control). I'm fine with `fail!` (but I think the bang is unnecessary as we don't have a non-bang version of it) itself.

**Comment by pascalbetz** [comment]
`fail` is the same as `throw`, so we need the bang.

**Comment by mereghost** [comment]
Not sure that I follow. 

Bang methods are usually destructive on the caller (and usually have a non-bang versions) and are usually meant to denote that they are more dangerous than their non-bang counterparts (i.e. `Kernel#exit` and `Kernel#exit!`) . If we need a bang because it interrupt the normal flow, so `Kernel#fail` would need one too.

`error` and `error!` have nothing in common at all, which kinda increases the confusion.

I know that's a breaking change, so probably not happening but random bang-methods is one of my pet peeves.

**Comment by jodosha** [comment]
@mereghost But without the bang, we're overriding `Kernel#fail`, which is dangerous. That's why in the first place I named it `#fail!` (with the bang).

**Comment by jodosha** [comment]
@pascalbetz  I'd like to have a decision about this soon.

  * Which problem are we solving?
  * Why the current implementation is not enough?
  * Which changes we should introduce to solve the problem?

Thank you! 😉 👍  /cc @mereghost

**Comment by mereghost** [comment]
@jodosha I thought that was the reason when I first saw it and I can't even think of a better name (naming is hard!) so we can remove the bang. 

But on the topic of `error` and `error!`: they have different semantics and differ only by the bang.

**Comment by pascalbetz** [comment]
@jodosha 

A bit difficult. It's more a "i'd do it differently" but have not worked extensively with Interactor, tbh.

I'd remove `error!` and provide a `fail!(msg = nil)` method instead to avoid the confusion between `error` and `error!` (remember, i started this because i was not aware of how to end the interactor with an error message).

**Comment by jodosha** [comment]
@pascalbetz fine 😄  - Please create a PR if you want. 😉  /cc @mereghost

**Comment by jodosha** [comment]
Closing as stale proposal. Feel free to act on this if you want.
